File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Effect/Combat/SmokeBlurEffect.ts
import { FairyGUI, UnityEngine } from "csharp";
import {AbstractEffect} from "mods/ModTheSpire/Scripts/Effect/AbstractEffect";
import UIHelper from "mods/ModTheSpire/Scripts/UI/UIHelper";
import FileHelper from "mods/ModTheSpire/Scripts/FileHelper";
import {Mathf} from "Core/Module/Math/Mathf";
import Color from "mods/ModTheSpire/Scripts/DataDefine/Color";
import {TimeKit} from "Core/Utils/TimeKit";

export class SmokeBlurEffect extends AbstractEffect{
    public Loader:FairyGUI.GLoader;
    private _Interval:number;
    private _TargetScale:number;
    private _Va:number;
    private _Vy:number;
    public constructor(public X: number,public Y:number) {
        super();
        this.Loader = UIHelper.CreateGLoader();
        if(Mathf.Random() < 0.5){
            this.Loader.url = FileHelper.FormatPath(`Vfx/exhaust/bigBlur.png`)
            this._TargetScale = Mathf.RandomRange(0.8,2.2);
        }else{
            this.Loader.url = FileHelper.FormatPath(`Vfx/exhaust/smallBlur.png`)
            this._TargetScale = Mathf.RandomRange(0.8,1.2);
        }
        this.Loader.SetScale(0.01,0.01);
        this.Loader.SetPosition(X+Mathf.RandomRange(-180,150),Y+Mathf.RandomRange(-240,150),0);
        let color = new Color();
        color.R = Mathf.RandomRange(128,153)
        color.G = color.R + Mathf.RandomRange(0, 51);
        color.B = 51;
        this.Loader.color = color.UnityColor();
        this._Interval = Mathf.Floor(Mathf.RandomRange(2,2.5));
        this.Loader.rotation = Mathf.RandomRange(0,360);
        this._Vy = Mathf.RandomRange(1,5);
        this._Va = Mathf.RandomRange(-250,250);
        this.Loader.TweenScale(new UnityEngine.Vector2(this._TargetScale,this._TargetScale),this._Interval).SetEase(FairyGUI.EaseType.ExpoOut);
        this.Loader.TweenFade(0,0.33).SetDelay(this._Interval - 0.33).SetEase(FairyGUI.EaseType.Linear);
    }
    
    Update() {
        super.Update();
        if(this.Duration >= this._Interval) {
            this.IsDone = true;
            this.Loader.Dispose();
            return;
        }
        this.Loader.x += Mathf.RandomRange(-2,2) + this._Vy;
        this.Loader.y += Mathf.RandomRange(-2,2) + this._Vy;
        this.Loader.rotation += this._Va * TimeKit.DeltaTime;
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Effect/Combat/SmokeBlurEffect.cs
using FairyGUI;
using UnityEngine;

namespace ModTheSpire.Effect.Combat
{
    public class SmokeBlurEffect : AbstractEffect
    {
        float mX;
        float mY;
        private float mVy;
        private float mVa;
        private GLoader mLoader;
        public SmokeBlurEffect(float x, float y)
        {
            mX = x;
            mY = y;
            
            mLoader = (GLoader)UIObjectFactory.NewObject(ObjectType.Loader);
            var targetScale = 0f;
            if(Random.value < 0.5){
                mLoader.url = ImageMaster.EXHAUST_L;
                targetScale = Random.Range(0.8F,2.2F);
            }else{
                mLoader.url = ImageMaster.EXHAUST_S;
                targetScale = Random.Range(0.8F,1.2F);
            }
            mLoader.SetScale(0.01f,0.01f);
            mLoader.SetPosition(x + Random.Range(-180, 150), y + Random.Range(-240, 150), 0);
            var color = new Color32();
            color.r = (byte)Random.Range(128, 153);
            color.g = (byte)(color.r + Random.Range(0, 51));
            color.b = 51;
            mLoader.color = color;
            var time = Random.Range(2, 2.5f);
            mLoader.rotation =Random.Range(0,360);
            this.mVy = Random.Range(1,5);
            this.mVa = Random.Range(-250,250);
            mLoader.TweenScale(new Vector2(targetScale,targetScale),time).SetEase(EaseType.ExpoOut).OnComplete(Finish);
            mLoader.TweenFade(0,0.33f).SetDelay(time - 0.33f).SetEase(EaseType.Linear);
        }

        public override void Update()
        {
            base.Update();
            mLoader.x += Random.Range(-2, 2) + mVy;
            mLoader.y += Random.Range(-2, 2) + mVy;
            mLoader.rotation += mVa * Time.deltaTime;
        }

        public override void Dispose()
        {
            base.Dispose();
            mLoader.Dispose();
        }
    }
}